// react-time-select/dist/TimeSelect.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var Input = require('react-bootstrap').Input;
var ReactIntl = require('react-intl');

var TimePicker = React.createClass({
  displayName: 'TimePicker',

  mixins: [ReactIntl.IntlMixin],

  propTypes: {
    className: React.PropTypes.string,
    label: React.PropTypes.string,
    name: React.PropTypes.string,
    value: React.PropTypes.instanceOf(Date),
    start: React.PropTypes.number,
    end: React.PropTypes.number,
    step: React.PropTypes.number,
    locale: React.PropTypes.string,
    onChange: React.PropTypes.func,
    formats: React.PropTypes.shape({
      time: React.PropTypes.shape({
        short: React.PropTypes.object
      })
    })
  },

  getDefaultProps: function getDefaultProps() {
    return {
      className: 'input-sm',
      label: 'Time',
      name: 'Time',
      // 00:30, 0030 is octal literal 24 & disallowed in strict mode.
      start: 30,
      end: 2359,
      step: 30,
      locale: 'en-GB',
      formats: {
        time: {
          short: {
            hour: '2-digit',
            minute: '2-digit'
          }
        }
      }
    };
  },

  onChange: function onChange(e) {
    if (this.props.onChange) {
      this.props.onChange(this.generateDateAtTime(e.target.value));
    }
  },

  listTimeOptions: function listTimeOptions() {
    var self = this;

    return this.generateTimeRange().map(function (unformattedTime) {
      var formattedTime = self.generateFormattedTime(unformattedTime);

      return {
        key: unformattedTime,
        value: formattedTime,
        date: self.generateDateAtTime(formattedTime)
      };
    });
  },

  generateFormattedTime: function generateFormattedTime(unformattedTime) {
    var formattedTime = '' + unformattedTime;

    while (formattedTime.length < 4) {
      formattedTime = '0' + formattedTime;
    }

    return formattedTime.replace(/(.{2})(.{2})/, '$1:$2');
  },

  generateDateAtTime: function generateDateAtTime(formattedTime) {
    var hoursMins = formattedTime.split(':');
    var newDate = new Date(this.props.value || new Date());
    newDate.setHours(hoursMins[0], hoursMins[1], 0, 0);

    return newDate;
  },

  generateTimeRange: function generateTimeRange() {
    var times = [];

    var start = parseInt(this.props.start, 10);
    var end = parseInt(this.props.end, 10);
    var step = parseInt(this.props.step, 10);

    for (var i = start; i < end; i += step) {
      var minutes = i % 100;

      if (minutes > 59) {
        i -= minutes;
        i += 100;
      }

      times.push(i);
    }

    return times;
  },

  defaultValueFromProps: function defaultValueFromProps() {
    if (!this.props.value) return undefined;

    return this.generateFormattedTime(this.props.value.getHours() * 100 + this.props.value.getMinutes());
  },

  render: function render() {
    var _this = this;

    return React.createElement(
      ReactIntl.IntlProvider,
      { locale: this.props.locale },
      React.createElement(
        Input,
        { type: 'select', value: this.defaultValueFromProps(), name: this.props.name, className: this.props.className, label: this.props.label, onChange: this.onChange },
        this.listTimeOptions().map(function (timeData) {
          return React.createElement(
            ReactIntl.FormattedTime,
            _extends({ key: timeData.key, value: timeData.date }, _this.props.formats.time.short),
            function (time) {
              return React.createElement(
                'option',
                { value: timeData.value },
                time
              );
            }
          );
        })
      )
    );
  }
});

module.exports = TimePicker;
